frontend/catalyst/utils/jax_extras.py

```diff
@@ -383,3 +401,80 @@ def new_inner_tracer(trace: DynamicJaxprTrace, aval) -> DynamicJaxprTracer:
     trace.frame.tracers.append(dt)
     trace.frame.tracer_to_var[id(dt)] = trace.frame.newvar(aval)
     return dt
+
+
+def get_implicit_and_explicit_flat_args(abstracted_axes, *args, **kwargs):
+    """Get implicit arguments from explicit arguments and abstracted_axes."""
+    axes_specs = _flat_axes_specs(abstracted_axes, *args, **kwargs)
+    explicit_args, _ = tree_flatten(args)
+    in_type = infer_lambda_input_type(axes_specs, explicit_args)
+    implicit_args = _extract_implicit_args(in_type, explicit_args)
+    args_flat = [*implicit_args, *explicit_args]
+    return args_flat
+
+
+def jaxpr_remove_implicit(
+    closed_jaxpr: ClosedJaxpr, out_type: OutputType
+) -> tuple[ClosedJaxpr, OutputType]:
+    """Remove all the implicit result values of the ``closed_jaxpr``."""
+    # Note: a more idiomatic way of doing this would be to re-trace the jaxpr and drop the unneeded
+    # tracers.
```

Comment:
This is interesting. I am a bit worried about changing this invariant that appears in JAX's code base.

Reply:
As discussed during the meeting, performance might be a big one, i.e. avoiding unnecessary walks through the entire program could be impactful.